### ISSUE TITLE
wasteful encoding of encrypted attachments in mime

### ISSUE THREAD

**Issue by tomholub** [FlowCrypt/flowcrypt-browser#4283]
When sending password-protected message, currently, we armor the encrypted data and then we base64 encode it.

Armoring also includes base64, and so we double-base64 encode it in the end.

It would be better to only encode once, if we can still produce a validly formatted message that way. This is browser:

```
------sinikael-?=_1-16432713087950.5202698292190784
Content-Type: application/octet-stream; name=encrypted.asc
Content-Description: OpenPGP encrypted message
Content-Disposition: inline; filename=encrypted.asc
X-Attachment-Id: f_VAPSRsspZNAlSauwuyhZpynklUNAXd@flowcrypt
Content-Id: <f_VAPSRsspZNAlSauwuyhZpynklUNAXd@flowcrypt>
Content-Transfer-Encoding: base64

LS0tLS1CRUdJTiBQR1AgTUVTU0FHRS0tLS0tDQpWZXJzaW9uOiBGbG93Q3J5cHQgRW1haWwgRW5j
cnlwdGlvbiA4LjIuNA0KQ29tbWVudDogU2VhbWxlc3NseSBzZW5kIGFuZCByZWNlaXZlIGVuY3J5
cHRlZCBlbWFpbA0KDQp3VjRERTRuUDM4SWwwaDRTQVFkQWN5eFJlbndqd1llTUR5dXNNMk9FeWhy
L0tkTTBMQ0s0ZlUwaE44cTcNCjQya3dvaGwwN0tLQ2tKV0tZM1VXSnQ4ZWM1TUh0Wld5SFdWR1Ez
eW9hWlB0VEkzN1FMTFJxSi82V3FxcA0KWmUrdHlkK0kwc0ppQVNpZktHZ2JUeFBMa1J5SkdsdDEz
REV1c05NYmlUM3d4azZjdS9oWjdJeGsrUU1uDQozY0t1QXllQXJHMUZZbkVyd3VFUlU1dHhMaTBQ
cjRxdDNTa2RvWWU2aGQzYTBGSEEwdmlGcTJYcnJYTjkNCjVERFdUanFoanIzU09mMm9NdXlsTm4y
WWZEY2FoZ3FJbzdkbVA3STVMYVFBUHRsTHhHRytLNmk3RFpSWA0Ka3VKZXBBQ3J4V1BjRmhRV2lv
UEw2VVRnS2NLMzRWWjNnWk1oMUNzWFRYMi90S3NLRGtxbjlwTHVhV1JxDQpqUnUzL0t2MkVXMms4
TVVJZ3BWRXpydTMzQ2cvT1NmM09VZDdQV2c0eXFIZkN2YmI3ZEl3US9UYzlvOTcNCkxhVmlVeDJB
c0VEbXMrd3NhaXJ2OElWQkRnbFp1WmF4bFI1THR2K3pMTGtnL3Azd2pQcGQvYlFUYmxjeA0KZXBY
a0RzQlJJZzV6enF5VlhYY1BCVDMwMlJ1QkxCaEcvVXJLczgraUZJU0lQbWFNYjd5a0dmcFRSVC9C
DQpZOUFHL09rNEhUN2praWdUQjY2Zk9qVlJXQmVDVUl3UldBYmY5aG5USFFKSXlNL3BwYnRaN0x2
cWFlTmUNCk5DS0kvY25HdUpadHlhQWN2QVJkc3BtT1l1RDNuSWtaTERoaDNmTnFDd3RRR2FNcW9u
OTRYVlp3dU5OOA0KQVExaG8wU1hoSjVtU2t3bFZjQ2taYWxJKzRQZ0xmK2hkdWp4WXVMUFdlQWhs
K2dZelJiS0p5R1RpZkQ5DQpDZFZpM3pUZXpqaG5iRERhK0dqMEJmQk56S0wxQ2pMOEtkeDFsa0ZT
S1lMNVp2MndDTnBGMHhkNCs5NmwNCkRJdzZDU0FIa3VkWXMvK054WTdUbnQ5SFlBNS9rWWEzU0hw
eHlwN3I1OGwvWEVwdDlIY0pDRUVlRGp2Tg0KcGRySVJOV3VGWnJncUUvL2RpamFtTUd2eXhFN3Yy
U2pmSFBhOGdTOTZWTWRhZ1M4b1pJalJWcVpiOXFqDQpKUmdVNE92LzZDazNIQXQ3bCtDVWc5QUZD
ZTA0Q2ViSHFPc01nL09nVmt5bGdHRkFETm9DdUg0UFZJLzUNCjdmWWtMK3N4dnBhTVp1aE1PSE1U
enFBUUpqMEFtazYxdDR2dm1TcTc4SWFjNWRTN3RPSGRYeWdCSGJYbw0KRCsyNDZGbk9xdjJZakMr
a1lFSHBJVkFJOE4xdnZoU2JldkZ2N3g2WVh1ekl6bXNkTTBwb2xVdzBZYTliDQpYSzVIWXg5bkFF
b0VkemNyS1RlZ1hHR1c4OTMwWGx1bnBsUGZlY0t0T2RhcnphdWRWd3k1dEVBVnVabTYNCkpHZ1M3
c3lwZXBQVlV2VEMrUWZVdGZ1c3ZIKzN1K2JGaXV3QlFJNDlBaStKNmx0cFlTSEFQOVI3cnpicA0K
RVE9PQ0KPVJsd3UNCi0tLS0tRU5EIFBHUCBNRVNTQUdFLS0tLS0NCg==
------sinikael-?=_1-16432713087950.5202698292190784--
```

Android does it better (7bit used here)

```
------=_Part_10_1347953.1643274037349
Content-Type: application/pgp-encrypted; name=message.asc
Content-Transfer-Encoding: 7bit
Content-Description: OpenPGP encrypted message
Content-Disposition: inline; filename=message.asc

-----BEGIN PGP MESSAGE-----
Version: PGPainless

hF4DE4nP38Il0h4SAQdAqcd3EtTvtyrPpFlx10l+fRWKJB5yE5bd0AkIjx2pdHEw
krUGVlQBcrrifbE5AFzOt2Y4fLT8MpsIGuYTsRSXbZSfVxoiZ3INI4bCvPBtwOyC
0sCXAcNs+sP/SqTjHbmYDFqN+zrf5NkJqUbe0t6wOWtGzIEPkS9gXGzXivzWRL3e
Kauf+Yz0OKSFIbjdEsb7dFdrZUUoDnyQfjvMHr77f+Iw1utjrbxO/uf1hEcorCKY
0wK+nJdU6ha33JrtcfKuURvPYLRdgtHW6v3ZbUOkxQeMM9MUhGBMSu5gfbOEGNOW
sUYrxDPXjLjIqFj96TLXggp7ShVXRGQE4kO0etC8d9XENw/d++3aHnW8IFsXpNY7
hC0TAylGzjmo6sOGEk0IK0PfaH93n1zzTxKJm/9yBPtvjqnI5C16tDdxQS5JtxTu
q3iXyOtYhrj8idXGl83pgrbbGaTdmdx4+oiFeSDgMl36PBXpuaoQcHHp345hczWb
kfKIdM1YmKpHhjR/NLAzxGCHJDGHs/zACtuwNH2TVwlOjAU+UHAtDs+uEHky/Emt
fVxsBjA2/gVZ2A==
=/7dw
-----END PGP MESSAGE-----

------=_Part_10_1347953.1643274037349--
```

**Comment by tomholub** [comment]
Also, we would not have to armor the attachments? We could just use binary, right? Please look into what would be the most space-efficient way to encode these (without compression), as email has limited size for attachments, and it would be best to utilize it to the maximum. Then, also double check what is the largest attachment size we can still safely send without triggering email oversize error on gmail.

**Comment by sosnovsky** [comment]
After checking PR https://github.com/FlowCrypt/flowcrypt-browser/pull/4951 I noticed that we currently don't include attachments in password-protected message (they're available only when opening on FES). But attachments are included when password-protected message is sent from iOS or Android app.

Then I found issue for non-included attachments - https://github.com/FlowCrypt/flowcrypt-browser/issues/4702, which is blocked by https://github.com/FlowCrypt/flowcrypt-browser/issues/2538 and this one probably can be fixed by https://github.com/FlowCrypt/flowcrypt-browser/issues/3032 (as noted here https://github.com/FlowCrypt/flowcrypt-browser/issues/2538#issuecomment-714299444)

And then we have another issue for wasteful encoding of encrypted attachments - https://github.com/FlowCrypt/flowcrypt-browser/issues/4041

I think it'll be better to fix encryption for all pgp/mime messages here (not only password-protected) and then re-enable attachments in password-protected messages.

@tomholub @rrrooommmaaa what do you think?

PS. @tomholub is there any update on https://partnerissuetracker.corp.google.com/issues/171715824 from google (filed in https://github.com/FlowCrypt/flowcrypt-browser/issues/3032)?

**Comment by rrrooommmaaa** [comment]
We can think of using a single format of outgoing messages. Take Android version as the base, for example. Is it using Gmail HTTP calls or some in-built OS functions?
It seems Android's version isn't using `multipart/alternative` while the browser version does, is this correct?
Gmail badly reacts to `multipart/alternative`

**Comment by rrrooommmaaa** [comment]
Android simply doesn't generate `text/html` version of `... has sent you a password-encrypted email.` so the message without `multipart/alternative` isn't modified in transit by gmail.

**Comment by sosnovsky** [comment]
> We can think of using a single format of outgoing messages. Take Android version as the base, for example. Is it using Gmail HTTP calls or some in-built OS functions?

It seems android uses `gmail` lib for sending messages - https://github.com/FlowCrypt/flowcrypt-android/blob/master/FlowCrypt/src/main/java/com/flowcrypt/email/api/email/gmail/GmailApiHelper.kt#L690, maybe it also reformats message before uploading to google

> Android simply doesn't generate `text/html` version of `... has sent you a password-encrypted email.` so the message without `multipart/alternative` isn't modified in transit by gmail.

Then we can try to use the same structure in browser version and check if it works well here too

**Comment by rrrooommmaaa** [comment]
> It seems android uses `gmail` lib for sending messages

This lib may use HTTP API calls too.
We also have `googleapis` in JS, but we're not using it in the extension code to save some space, I suppose.

> Then we can try to use the same structure in browser version and check if it works well here too

It is useful indeed. In a separate issue, I guess.

**Comment by sosnovsky** [comment]
Okay, it seems we can remove `base64` encoding of message only by deleting `multipart/alternative` and using just `text/plain` or `text/html`. And it won't give us large message size decrease, as we're currently not including attachments for password-protected messages.

I'll merge current PR, so we'll start using `7bit` for content transfer encoding. And then the next step will be fixing https://github.com/FlowCrypt/flowcrypt-browser/issues/4041, which will allow us to re-enable attachments for password protected messages (https://github.com/FlowCrypt/flowcrypt-browser/issues/4702)
